### PR TITLE
Enforce product SKU uniqueness at the database layer.

### PR DESCRIPTION
**Pull request by akaplya**
Enforce product SKU uniqueness at the database layer.

### CODE REVIEW

**Comment by DavidLambauer** [comment]
It would be helpful to shortly describe the current mechanism.

**Comment by buskamuza** [comment]
https://www.practicalecommerce.com/Inventory-Labeling-Basics-SKUs-UPCs-EANs

File: design-documents/catalog/database/product-sku-unique-field.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+##Make product SKU a unique field in a product table.
+For now, responsibility for the uniqueness of SKU attribute is part of application logic.
+As a result, this logic may be inconsistent per product creation scenarios.
+For instance, import may have another set of validations rules than product creation through the admin panel.
+
+To ensure uniqueness of product SKU in the Magento database, we should add a unique constraint for the SKU field in the product table.
+
+For backward compatibility, we can enforce SKU uniqueness through the database patch. Product with a higher ID will keep an SKU. All duplicates will get a unique suffix. For transparency, the merchant will get a report of duplicates that were resolved by the patch.
```
Comment (Vinai) [inline design-documents/catalog/database/product-sku-unique-field.md]: A dry-run option to generate the report before applying it would be really helpful.

**Comment by akaplya** [comment]
we can not enforce unique constraint for SKU because we will not be able to support it for Enterprise edition

**Pull request closed without merging**